File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {

        PriorityQueue<int[]> queue = new PriorityQueue<int[]>((p1, p2) -> {

            int dis1 = p1[0]*p1[0] + p1[1]*p1[1];
            int dis2 = p2[0]*p2[0] + p2[1]*p2[1];           

            return dis2 - dis1;
        });
        
        for(int[] point : points){
            queue.offer(point);

            if (queue.size() > k){
                queue.poll();
            }
        }

        return queue.toArray(new int[0][0]);
    }
}

File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        
        #Time Complexity : O(N*log(K))
        #Space Comelexity : O(N)

        maxHeap = []
        for point in points:
            distance = point[0]**2 + point[1]**2
            heapq.heappush(maxHeap, (-distance, point))
            if len(maxHeap) > k:
                heapq.heappop(maxHeap)
        res = []
        for value in maxHeap:
            res.append(value[1])
        return res